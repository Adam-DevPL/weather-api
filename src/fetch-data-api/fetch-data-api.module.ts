import { HttpModule } from '@nestjs/axios';
import { Module } from '@nestjs/common';
import { FetchDataApiFilter } from 'src/filters/fetch-data-api.filter';
import { FetchDataApiService } from './fetch-data-api.service';
import { APP_FILTER } from '@nestjs/core';

@Module({
  imports: [HttpModule],
  providers: [
    FetchDataApiService,
    {
      provide: APP_FILTER,
      useClass: FetchDataApiFilter,
    },
  ],
  exports: [FetchDataApiService],
})
export class FetchDataApiModule {}
