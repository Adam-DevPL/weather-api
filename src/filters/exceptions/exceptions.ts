export class LocationException extends Error {
  constructor(readonly message: string) {
    super(message);
    this.name = 'LocationException';
  }
}

export class CoordinatesException extends Error {
  constructor(readonly message: string) {
    super(message);
    this.name = 'CoordinatesException';
  }
}

export class InternalServerException extends Error {
  constructor(readonly message: string) {
    super(message);
    this.name = 'InternalServerException';
  }
}

export class LocationTypeException extends Error {
  constructor(readonly message: string) {
    super(message);
    this.name = 'LocationTypeException';
  }
}
