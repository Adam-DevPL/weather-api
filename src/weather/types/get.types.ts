import { GeoCoordinates } from 'src/weather/types/weather.types';

export type ApiResponse = {
  avgTemperature: number;
  weather: string;
};

export type WeatherResponse = {
  location: string | GeoCoordinates;
  avgTemperature: number;
  weather: string;
};
