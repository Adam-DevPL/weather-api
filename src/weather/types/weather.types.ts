type DateInput = {
  day: string;
};
type DatesRange = {
  from: string;
  to: string;
};

export type GeoCoordinates = {
  latitude: number;
  longitude: number;
};

export type LocationNameInput = {
  locationType: LocationType.CITY | LocationType.COUNTRY;
  locationName: string;
};

export type LocationNameInputDay = LocationNameInput & DateInput;
export type LocationNameInputDates = LocationNameInput & DatesRange;

export type CoordinatesInput = {
  lat: number;
  lon: number;
};

export type CoordinatesInputDates = CoordinatesInput & DatesRange;

export enum LocationType {
  CITY,
  COUNTRY,
  GEO,
}
