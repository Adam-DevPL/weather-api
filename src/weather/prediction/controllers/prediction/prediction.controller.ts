import {
  Controller,
  Get,
  HttpException,
  HttpStatus,
  Param,
} from '@nestjs/common';
import { PredictionService } from 'src/prediction/services/prediction/prediction.service';
import { ForecastResponse } from 'src/prediction/types/prediction.types';
import { ToolsService } from 'src/tools/tools.service';
import {
  CoordinatesInputDates,
  LocationNameInputDates,
  LocationNameInputDay,
  LocationType,
} from 'src/weather/types/weather.types';
import {
  CityDatesRangeParam,
  CityDayParam,
  CountryDatesRangeParam,
  CountryDayParam,
  GeoCoordinatesDatesRangeParam,
} from 'src/Validation/validation.module';

@Controller('weather/prediction')
export class PredictionController {
  constructor(
    private predictionService: PredictionService,
    private toolsService: ToolsService,
  ) {}

  @Get('country/:day/:country')
  getForecastForCountry(
    @Param() { country, day }: CountryDayParam,
  ): Promise<ForecastResponse> {
    const locationNameInputDay: LocationNameInputDay = {
      day,
      locationName: country,
      locationType: LocationType.COUNTRY,
    };

    return this.predictionService.getForecastForSingleDay(locationNameInputDay);
  }
  @Get('city/:day/:city')
  getForecastForCity(
    @Param() { city, day }: CityDayParam,
  ): Promise<ForecastResponse> {
    const locationNameInputDay: LocationNameInputDay = {
      day,
      locationName: city,
      locationType: LocationType.CITY,
    };

    return this.predictionService.getForecastForSingleDay(locationNameInputDay);
  }

  @Get('country/:country/:from/:to')
  getForecastForCountryInDateRange(
    @Param() { country, from, to }: CountryDatesRangeParam,
  ): Promise<ForecastResponse> {
    if (this.toolsService.checkDatesRange(from, to)) {
      throw new HttpException(
        'Invalid dates order. Date "from" is later then "to" date',
        HttpStatus.BAD_REQUEST,
      );
    }
    const locationNameInputDates: LocationNameInputDates = {
      locationType: LocationType.COUNTRY,
      locationName: country,
      from,
      to,
    };

    return this.predictionService.getForecastForCountryOrCityInDateRange(
      locationNameInputDates,
    );
  }

  @Get('city/:city/:from/:to')
  getForecastForCityInDateRange(
    @Param() { city, from, to }: CityDatesRangeParam,
  ): Promise<ForecastResponse> {
    if (this.toolsService.checkDatesRange(from, to)) {
      throw new HttpException(
        'Invalid dates order. Date "from" is later then "to" date',
        HttpStatus.BAD_REQUEST,
      );
    }

    const locationNameInputDates: LocationNameInputDates = {
      locationType: LocationType.CITY,
      locationName: city,
      from,
      to,
    };

    return this.predictionService.getForecastForCountryOrCityInDateRange(
      locationNameInputDates,
    );
  }

  @Get('location/:lat/:lon/:from/:to')
  getForecastForCoordinatesInDateRange(
    @Param() { lat, lon, from, to }: GeoCoordinatesDatesRangeParam,
  ): Promise<ForecastResponse> {
    if (this.toolsService.checkDatesRange(from, to)) {
      throw new HttpException(
        'Invalid dates order. Date "from" is later then "to" date',
        HttpStatus.BAD_REQUEST,
      );
    }
    const coordinatesInputDates: CoordinatesInputDates = {
      lat,
      lon,
      from,
      to,
    };

    return this.predictionService.getForecastForGeoLocationInDateRange(
      coordinatesInputDates,
    );
  }
}
