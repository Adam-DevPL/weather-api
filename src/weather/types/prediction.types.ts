import { GeoCoordinates } from 'src/weather/types/weather.types';

export type ForecastResponse = {
  location: string | GeoCoordinates;
  weatherInfo: WeatherInfo | WeatherInfo[];
};

export type WeatherInfo = {
  day: string;
  avgTemperature: string;
  weather: string;
};

export type OpenModelResponse = {
  data: {
    hourly: {
      time: string[];
      temperature_2m: number[];
      weathercode: number[];
    };
  };
};

export type ApiResponsePredictRoute = {
  timeArr: string[];
  temperatureArr: number[];
  weatherCodeArr: number[];
};
