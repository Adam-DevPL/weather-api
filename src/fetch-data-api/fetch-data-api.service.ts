import { HttpService } from '@nestjs/axios';
import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

import { firstValueFrom } from 'rxjs';
import { LocationType } from 'src/prediction/types/prediction.types';
import {
  FetchDataApiGeoResponse,
  FetchDataApiParams,
} from './types/FetchDataApi.types';

@Injectable()
export class FetchDataApiService {
  private readonly weatherUrl;
  private readonly geoUrl;

  constructor(
    private readonly httpService: HttpService,
    private readonly configService: ConfigService,
  ) {
    this.weatherUrl = this.configService.get<string>('WEATHER_URL');
    this.geoUrl = this.configService.get<string>('GEO_URL');
  }

  public async getDataFromApi(params: FetchDataApiParams) {
    try {
      const { data } = await firstValueFrom(
        this.httpService.get(this.weatherUrl, { params }),
      );
      if (data.error) {
        throw new HttpException(
          { status: HttpStatus.BAD_REQUEST, message: 'Invalid coordinates' },
          HttpStatus.BAD_REQUEST,
        );
      }

      return {
        data,
      };
    } catch ({ response }) {
      throw new HttpException(
        {
          status:
            response.status === 404
              ? HttpStatus.INTERNAL_SERVER_ERROR
              : HttpStatus.BAD_REQUEST,
          error:
            response.status === 404
              ? 'Internal server error'
              : response.message,
        },
        response.status === 404
          ? HttpStatus.INTERNAL_SERVER_ERROR
          : HttpStatus.BAD_REQUEST,
        {
          cause: response,
        },
      );
    }
  }

  public async getGeoLocation(name: string): Promise<FetchDataApiGeoResponse> {
    try {
      const params = {
        name,
        count: 1,
      };

      const { data } = await firstValueFrom(
        this.httpService.get(this.geoUrl, { params }),
      );

      if (!data.hasOwnProperty('results')) {
        throw new HttpException(
          { status: HttpStatus.BAD_REQUEST, message: 'Location not found' },
          HttpStatus.BAD_REQUEST,
        );
      }

      const {
        results: [{ latitude, longitude, country }],
      } = data;

      return {
        latitude,
        longitude,
        location:
          country.toLowerCase() === name.toLowerCase()
            ? LocationType.Country
            : LocationType.City,
      };
    } catch ({ response }) {
      throw new HttpException(
        {
          status:
            response.status === 404
              ? HttpStatus.INTERNAL_SERVER_ERROR
              : HttpStatus.BAD_REQUEST,
          error:
            response.status === 404
              ? 'Internal server error'
              : response.message,
        },
        response.status === 404
          ? HttpStatus.INTERNAL_SERVER_ERROR
          : HttpStatus.BAD_REQUEST,
        {
          cause: response,
        },
      );
    }
  }
}
