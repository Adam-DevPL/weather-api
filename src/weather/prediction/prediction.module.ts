import { Module } from '@nestjs/common';
import { FetchDataApiModule } from 'src/fetch-data-api/fetch-data-api.module';
import { ToolsModule } from 'src/tools/tools.module';
import { PredictionController } from './controllers/prediction.controller';
import { PredictionService } from './services/prediction.service';

@Module({
  imports: [FetchDataApiModule, ToolsModule],
  controllers: [PredictionController],
  providers: [PredictionService],
})
export class PredictionModule {}
