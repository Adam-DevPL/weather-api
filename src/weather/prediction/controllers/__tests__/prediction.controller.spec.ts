import { Test, TestingModule } from '@nestjs/testing';
import { PredictionController } from '../prediction.controller';
import { ConfigService } from '@nestjs/config';
import { HttpModule } from '@nestjs/axios';
import { FetchDataApiService } from 'src/fetch-data-api/fetch-data-api.service';
import { ToolsService } from 'src/tools/tools.service';
import { HttpException, HttpStatus } from '@nestjs/common';
import { LocationType } from 'src/weather/types/weather.types';
import {
  CityDayParam,
  CountryDatesRangeParam,
  CountryDayParam,
  GeoCoordinatesDatesRangeParam,
} from 'src/validation/validation.module';
import { plainToInstance } from 'class-transformer';
import { validate } from 'class-validator';
import { PredictionService } from '../../services/prediction.service';
import { ForecastResponse } from 'src/weather/types/prediction.types';
import { LocationTypeException } from 'src/filters/exceptions/exceptions';

describe('PredictionController', () => {
  let controller: PredictionController;
  let service: PredictionService;
  let fetchDataApi: FetchDataApiService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [HttpModule],
      controllers: [PredictionController],
      providers: [
        PredictionService,
        FetchDataApiService,
        ToolsService,
        ConfigService,
      ],
    }).compile();

    controller = module.get<PredictionController>(PredictionController);
    service = module.get<PredictionService>(PredictionService);
    fetchDataApi = module.get<FetchDataApiService>(FetchDataApiService);
  });

  it('should be defined', () => {
    expect(controller).toBeDefined();
  });

  describe('getForecastForCountry', () => {
    it('should return proper weather response for a counry', async () => {
      //given
      const mockWeatherResponse: ForecastResponse = {
        location: 'country',
        weatherInfo: {
          day: '2023-02-18',
          avgTemperature: '5.5',
          weather: 'sunny',
        },
      };
      jest
        .spyOn(service, 'getForecastForSingleDay')
        .mockImplementation(() => Promise.resolve(mockWeatherResponse));

      //when
      const result = await controller.getForecastForCountry({
        country: 'country',
        day: '2023-02-18',
      });

      //then
      expect(result).toMatchSnapshot();
    });

    it('should throw an error when invalid country', async () => {
      //given
      jest.spyOn(service, 'getForecastForSingleDay').mockImplementation(() => {
        throw new LocationTypeException(
          `It is not a ${LocationType[LocationType.COUNTRY]}: invalidCountry`,
        );
      });

      await expect(
        controller.getForecastForCountry({
          day: '2023-02-22',
          country: 'Katowice',
        }),
      ).rejects.toMatchSnapshot();
    });

    it('should throw error, when country is not valid - not a string only with letters', async () => {
      //given
      const myBodyObjectNumber = { country: 123, day: '2023-02-18' };
      const myBodyObjectStringWithNonLetters = {
        country: 'Pol123!@',
        day: '2023-02-18',
      };
      const myObject1 = plainToInstance(CountryDayParam, myBodyObjectNumber);
      const myObject2 = plainToInstance(
        CountryDayParam,
        myBodyObjectStringWithNonLetters,
      );

      //when
      const errors1 = await validate(myObject1);
      const errors2 = await validate(myObject2);

      //then
      expect(errors1.length).not.toBe(0);
      expect(errors2.length).not.toBe(0);
      expect(JSON.stringify(errors1)).toContain(`country must be a string`);
      expect(JSON.stringify(errors2)).toContain(
        `country must contain only letters (a-zA-Z)`,
      );
    });

    it('should throw error, when day is before actual or after 7 days of current date', async () => {
      //given
      const myBodyObjectNumber = { country: 'Poland', day: '2023-02-10' };
      const myBodyObjectStringWithNonLetters = {
        country: 'Poland',
        day: '2023-03-27',
      };
      const myObject1 = plainToInstance(CountryDayParam, myBodyObjectNumber);
      const myObject2 = plainToInstance(
        CountryDayParam,
        myBodyObjectStringWithNonLetters,
      );

      //when
      const errors1 = await validate(myObject1);
      const errors2 = await validate(myObject2);

      //then
      expect(errors1.length).not.toBe(0);
      expect(errors2.length).not.toBe(0);
      expect(JSON.stringify(errors1)).toContain(
        `Date is out of range: min +1 day and max +7 days`,
      );
      expect(JSON.stringify(errors2)).toContain(
        `Date is out of range: min +1 day and max +7 days`,
      );
    });
  });

  describe('getForecastForCity', () => {
    it('should return proper weather response for a city', async () => {
      //given
      const mockWeatherResponse: ForecastResponse = {
        location: 'city',
        weatherInfo: {
          day: '2023-02-18',
          avgTemperature: '5.5',
          weather: 'sunny',
        },
      };
      jest
        .spyOn(service, 'getForecastForSingleDay')
        .mockImplementation(() => Promise.resolve(mockWeatherResponse));

      //when
      const result = await controller.getForecastForCity({
        city: 'city',
        day: '2023-02-18',
      });

      //then
      expect(result).toMatchSnapshot();
    });

    it('should throw an error when invalid city', async () => {
      //given
      jest.spyOn(service, 'getForecastForSingleDay').mockImplementation(() => {
        throw new LocationTypeException(
          `It is not a ${LocationType[LocationType.CITY]}: invalidCity`,
        );
      });

      await expect(
        controller.getForecastForCity({
          day: '2023-02-24',
          city: 'invalidCity',
        }),
      ).rejects.toMatchSnapshot();
    });

    it('should throw error, when city is not valid - not a string only with letters', async () => {
      //given
      const myBodyObjectNumber = { city: 123, day: '2023-02-18' };
      const myBodyObjectStringWithNonLetters = {
        city: 'Pol123!@',
        day: '2023-02-18',
      };
      const myObject1 = plainToInstance(CityDayParam, myBodyObjectNumber);
      const myObject2 = plainToInstance(
        CityDayParam,
        myBodyObjectStringWithNonLetters,
      );

      //when
      const errors1 = await validate(myObject1);
      const errors2 = await validate(myObject2);

      //then
      expect(errors1.length).not.toBe(0);
      expect(errors2.length).not.toBe(0);
      expect(JSON.stringify(errors1)).toContain(`city must be a string`);
      expect(JSON.stringify(errors2)).toContain(
        `city must contain only letters (a-zA-Z)`,
      );
    });

    it('should throw error, when day is before actual or after 7 days of current date', async () => {
      //given
      const myBodyObjectNumber = { city: 'Poland', day: '2023-02-10' };
      const myBodyObjectStringWithNonLetters = {
        city: 'Poland',
        day: '2023-03-27',
      };
      const myObject1 = plainToInstance(CityDayParam, myBodyObjectNumber);
      const myObject2 = plainToInstance(
        CityDayParam,
        myBodyObjectStringWithNonLetters,
      );

      //when
      const errors1 = await validate(myObject1);
      const errors2 = await validate(myObject2);

      //then
      expect(errors1.length).not.toBe(0);
      expect(errors2.length).not.toBe(0);
      expect(JSON.stringify(errors1)).toContain(
        `Date is out of range: min +1 day and max +7 days`,
      );
      expect(JSON.stringify(errors2)).toContain(
        `Date is out of range: min +1 day and max +7 days`,
      );
    });
  });

  describe('getForecastForCountryInDateRange', () => {
    it('should return proper weather response for a country', async () => {
      //given
      const mockWeatherResponse: ForecastResponse = {
        location: 'country',
        weatherInfo: [
          {
            day: '2023-02-18',
            avgTemperature: '5.5',
            weather: 'sunny',
          },
          {
            day: '2023-02-19',
            avgTemperature: '5.5',
            weather: 'sunny',
          },
          {
            day: '2023-02-20',
            avgTemperature: '5.5',
            weather: 'sunny',
          },
        ],
      };

      jest
        .spyOn(service, 'getForecastForCountryOrCityInDateRange')
        .mockImplementation(() => Promise.resolve(mockWeatherResponse));

      //when
      const result = await controller.getForecastForCountryInDateRange({
        country: 'country',
        from: '2023-02-15',
        to: '2023-02-17',
      });

      //then
      expect(result).toMatchSnapshot();
    });

    it('should throw an error when invalid country', async () => {
      //given
      jest
        .spyOn(service, 'getForecastForCountryOrCityInDateRange')
        .mockImplementation(() => {
          throw new LocationTypeException(
            `It is not a ${LocationType[LocationType.COUNTRY]}: invalidCountry`,
          );
        });

      await expect(
        controller.getForecastForCountryInDateRange({
          from: '2023-02-22',
          to: '2023-02-25',
          country: 'invalidCountry',
        }),
      ).rejects.toMatchSnapshot();
    });

    it('should throw error, when country is not valid - not a string only with letters', async () => {
      //given
      const myBodyObjectNumber = {
        country: 123,
        from: '2023-02-18',
        to: '2023-02-20',
      };
      const myBodyObjectStringWithNonLetters = {
        city: 'Pol123!@',
        from: '2023-02-18',
        to: '2023-02-20',
      };
      const myObject1 = plainToInstance(
        CountryDatesRangeParam,
        myBodyObjectNumber,
      );
      const myObject2 = plainToInstance(
        CountryDatesRangeParam,
        myBodyObjectStringWithNonLetters,
      );

      //when
      const errors1 = await validate(myObject1);
      const errors2 = await validate(myObject2);

      //then
      expect(errors1.length).not.toBe(0);
      expect(errors2.length).not.toBe(0);
      expect(JSON.stringify(errors1)).toContain(`country must be a string`);
      expect(JSON.stringify(errors2)).toContain(
        `country must contain only letters (a-zA-Z)`,
      );
    });

    it('should throw error, when day is before actual or after 7 days of current date', async () => {
      //given
      const myBodyObjectNumber = {
        country: 'Poland',
        from: '2023-02-10',
        to: '2023-02-27',
      };
      const myObject1 = plainToInstance(
        CountryDatesRangeParam,
        myBodyObjectNumber,
      );

      //when
      const errors = await validate(myObject1);

      //then
      expect(errors.length).not.toBe(0);
      expect(JSON.stringify(errors)).toContain(
        `Date is out of range: min +1 day and max +7 days`,
      );
    });

    it('should throw error, when from date is later then to date', async () => {
      //given
      // jest
      //   .spyOn(service, 'getForecastForCountryOrCityInDateRange')
      //   .mockImplementation(() => {
      //     throw new HttpException(
      //       `Invalid dates order. Date "from" is later then "to" date`,
      //       HttpStatus.BAD_REQUEST,
      //     );
      //   });

      await expect(
        controller.getForecastForCountryInDateRange({
          from: '2023-02-26',
          to: '2023-02-23',
          country: 'country',
        }),
      ).rejects.toMatchSnapshot();
    });
  });

  describe('getForecastForCityInDateRange', () => {
    //the same test as for country
  });

  describe('getForecastForCoordinatesInDateRange', () => {
    //the same test as for country + below
    it('should return error when latitude or longitude is out of range -180 < x < 180', async () => {
      //given
      const myBodyObject = {
        lat: -200,
        lon: 79,
        from: '2023-02-10',
        to: '2023-02-27',
      };
      const myObject = plainToInstance(
        GeoCoordinatesDatesRangeParam,
        myBodyObject,
      );

      //when
      const errors = await validate(myObject);

      //then
      expect(errors.length).not.toBe(0);
      expect(JSON.stringify(errors)).toContain(
        `Date is out of range: min +1 day and max +7 days`,
      );
      expect(JSON.stringify(errors)).toContain(
        `Latitude and longitude must be in range -180 < x < 180`,
      );
    });
  });
});
