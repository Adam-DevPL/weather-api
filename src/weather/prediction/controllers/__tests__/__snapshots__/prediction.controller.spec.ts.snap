// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PredictionController getForecastForCity should return proper weather response for a city 1`] = `
{
  "location": "city",
  "weatherInfo": {
    "avgTemperature": "5.5",
    "day": "2023-02-18",
    "weather": "sunny",
  },
}
`;

exports[`PredictionController getForecastForCountry should return proper weather response for a counry 1`] = `
{
  "location": "country",
  "weatherInfo": {
    "avgTemperature": "5.5",
    "day": "2023-02-18",
    "weather": "sunny",
  },
}
`;

exports[`PredictionController getForecastForCountry should throw an error when invalid country 1`] = `[InternalServerException: Invalid URL]`;

exports[`PredictionController getForecastForCountryInDateRange should return proper weather response for a country 1`] = `
{
  "location": "country",
  "weatherInfo": [
    {
      "avgTemperature": "5.5",
      "day": "2023-02-18",
      "weather": "sunny",
    },
    {
      "avgTemperature": "5.5",
      "day": "2023-02-19",
      "weather": "sunny",
    },
    {
      "avgTemperature": "5.5",
      "day": "2023-02-20",
      "weather": "sunny",
    },
  ],
}
`;

exports[`PredictionController getForecastForCountryInDateRange should throw error, when from date is later then to date 1`] = `[InternalServerException: Invalid URL]`;
